radix-engine/src/wasm/wasmer.rs

```diff
@@ -206,6 +216,11 @@ impl WasmerModule {
             let ident = read_memory(&instance, ident_ptr, ident_len)?;
             let args = read_memory(&instance, args_ptr, args_len)?;
 
+            // Get current memory consumption and set it in current call frame through api.
+            runtime
+                .set_wasm_memory_consumption(get_memory_size(&instance)?)
+                .map_err(|e| RuntimeError::user(Box::new(e)))?;
```

Comment:
We need to store/check what is the current wasm instance memory usage (call frame 1) before starting nested call to the next wasm instance (call frame 2). Without reading current memory before starting next call frame, we will get information of wasm consumed memory in `after_wasm_instantiation` event (for call frame 1) which will be called after nested call finishes (call frame 2). We will have the consumed memory information too late to properly validate the limit and break transaction execution.

Reply:
In that case, let's notify the transaction limits module in both `call_method` and `call_function`.

Reply:
Another point for future thinking - do we really need wasm memory usage limiting if we can't apply real-time WASM limiting?

Reply:
Current implementation is not real-time, but it can verify memory size in these two cases (before nested call and at invocation end) I think node can use this mechanism during preparation step and discard transaction before execution, maybe @malexandru-rdx can say something more about use of this limitation mechanism on node side.

I was looking to the wasm implementation and maybe we can use maximum number of memory pages which wasm `grow` function can extend. In wat files in `radix-engine-test` folder we are declaring memory like this: `(memory $0 1)`, but we can add third parameter to limit maximum count of pages which wasmi implementation (wasmer to check) validates on each call to `grow`: `(memory $0 1 5)` - this lets to extend memory up to 5 pages. I am not sure how we can add this mechanism to Scrypto compiled files, but maybe it is something worth to try.

Another thing which we can do is to create `grow` function instrumentation, similar to instrumentation of wasmi call stack height from `wasm-instrument` crate.

Reply:
In `enforce_memory_limit()` in prepare.rs file we are checking wasm pages limits, but only initial count, we are not checking maximum wasm pages count, so in this place we can add validation if that value is set and has proper value.

Reply:
That's an interesting idea and we should explore more.

If we inject a max memory size during preparation, we can have a nice upper bounds of memory usage per instance, and we don't need to track per-call-frame usage as well.

Reply:
As we discussed I've reverted max wasm memory limit (moved implementation to branch: `experimental/wasm_max_memory_limit`.